sdk/sdk-core/aar/src/main/java/mil/emp3/core/services/kml/KMLSRequest.java

```diff
@@ -0,0 +1,254 @@
+
+import android.content.Context;
+import android.util.Log;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.URLConnection;
+import java.util.UUID;
+
+import mil.emp3.api.exceptions.EMP_Exception;
+import mil.emp3.api.interfaces.IKML;
+import mil.emp3.api.interfaces.IKMLS;
+import mil.emp3.api.interfaces.IMap;
+import mil.emp3.api.interfaces.core.storage.IKMLSRequest;
+
+/**
+ * Holds the KML Request till it is removed via removeMapService.
+ *    1. Maintains the current state of the request.
+ *    2. Copies the file to application private folder.
+ */
+public class KMLSRequest implements KMZFile.IKMZFileRerquest, IKMLSRequest {
+    private static String TAG = KMLSRequest.class.getSimpleName();
+
+    private final int READ_BUFFER_SIZE = 4096;
+
+    private final String KMLS_ROOT = "KMLS";     // Root folder under which all KMLS service files are stored.
+
+    private static boolean directoryBuilt = false;
+    private static File kmlsRoot;
+
+    private IMap map;                           // Client on which the request was initiated
+    private IKMLS service;                      // KMLS service object built by the application.
+    private String kmzFilePath = null;          // KMZ file that was copied using user supplied URL (could be kml file also)
+    private File kmzDirectory = null;           // Directory where files for this request are stored.
+    private String kmlFilePath = null;          // Extracted kml file or file specified in the URL. This is the file that will be parsed.
+    private IKML feature = null;                       // Feature that was created by the parser.
+
+    KMLSRequest(IMap map, IKMLS service) {
+        this.map = map;
+        this.service = service;
+    }
+
+    @Override
+    public File getDestinationDir() {
+        return kmzDirectory;
+    }
+
+    @Override
+    public String getSourceFilePath() {
+        return kmzFilePath;
+    }
+
+    @Override
+    public String getKmlFilePath() {
+        return kmlFilePath;
+    }
+
+    @Override
+    public void setKmlFilePath(String kmlFilePath) {
+        this.kmlFilePath = kmlFilePath;
+    }
+
+    public IMap getMap() {
+        return map;
+    }
+
+    public void setMap(IMap map) {
+        this.map = map;
+    }
+
+    public IKMLS getService() {
+        return service;
+    }
+
+    public void setService(IKMLS service) {
+        this.service = service;
+    }
+
+    public File getKmzDirectory() { return kmzDirectory; }
+
+    public String getKmzFilePath() { return kmzFilePath; }
+
+    public IKML getFeature() { return feature; }
+
+    public void setFeature(IKML feature) { this.feature = feature; }
+
+    @Override
+    public UUID getId() {
+        if(null != getService()) {
+            return getService().getGeoId();
+        } else {
+            Log.e(TAG, "getId service is null");
+            return null;
```

Comment:
This not service having null geoId, this is checking if service itself is null. This should never happen. But I don't want to crash the APP.